---
title: "Dias/Koup preprocessing of RNA-Seq data (set 2)"
author: "Slim Fourati"
date: "`r Sys.Date()`"
output: github_document
---

Load packages
```{r load-packages}
suppressPackageStartupMessages(library(package = "knitr"))
suppressPackageStartupMessages(library(package = "kableExtra"))
suppressPackageStartupMessages(library(package = "readxl"))
suppressPackageStartupMessages(library(package = "EDASeq"))
suppressPackageStartupMessages(library(package = "pvca")) # dleelab/pvca
suppressPackageStartupMessages(library(package = "lme4"))
suppressPackageStartupMessages(library(package = "parallelDist"))
suppressPackageStartupMessages(library(package = "parallel"))
suppressPackageStartupMessages(library(package = "edgeR"))
suppressPackageStartupMessages(library(package = "ggbeeswarm"))
suppressPackageStartupMessages(library(package = "ComplexHeatmap"))
suppressPackageStartupMessages(library(package = "tidyverse"))
```

```{r session-options}
opts_chunk$set(echo = TRUE, fig.path = "../figure/")
options(readr.show_col_types   = FALSE,
        dplyr.summarise.inform = FALSE)
workDir <- dirname(getwd())
```

```{r load-raw-seqset, echo=FALSE}
load(file = file.path(workDir, "output/joana2.esetRaw.RData"))
countDF <- counts(esetRaw)
sampleAnnotDF <-pData(esetRaw)
featAnnot <- fData(esetRaw)
```

Reading raw counts
```{r read-counts, eval=FALSE}
countDF <- read_csv(file = file.path(workDir, "input/joana2.genecounts.csv")) %>%
  as.data.frame() %>%
  column_to_rownames(var = "gene_id")
```

```{r print-counts}
message("First five rows/columns of count matrix")
countDF[1:5, 1:5] %>%
  kable()
message("Dimensions of the count matrix")
print(dim(countDF))
```

Read sample annotation
```{r read-rnaseq-samplesheet, eval=FALSE}
sampleAnnotFile <- file.path(workDir,
                             "input",
                             "JoanaDias_Bulk_RNAseq_sampleTracking_06012021.xlsx")
# read each plates
sampleAnnotDF <- read_excel(path = sampleAnnotFile,
                            .name_repair = function(x) make.unique(make.names(x)))
# append treatment info
treatFile <- file.path(workDir,
                       "input",
                       "RNAseq_II_for_Slim_20210829.slide3.txt")
treatDF <- read_tsv(file = treatFile, name_repair = make.names)
sampleAnnotDF <- merge(x = sampleAnnotDF, y = treatDF, by = "Monkey.ID")
# append sample id
sampleAnnotDF <- sampleAnnotDF %>%
  mutate(rowname = paste(Monkey.ID,
                         Timepoint,
                         Cell.population.sorted.1,
                         sep = "_"),
         rowname = gsub(pattern = " ", replacement = "", rowname)) %>%
   as.data.frame() %>%
  column_to_rownames(var = "rowname")
sampleAnnotDF <- sampleAnnotDF[colnames(countDF), ]
# TotalReads counted
sampleAnnotDF$TotalReads <- colSums(countDF)
```

```{r print-rnaseq-samplesheet}
message("print header of sample annotation")
sampleAnnotDF[1:5, 1:5] %>%
  kable()
message("Number of animal w/ RNA-seq data")
sampleAnnotDF %>%
  .$Monkey.ID %>%
  unique() %>%
  length() %>%
  print()
message("Number of timepoints and subsets w/ RNA-seq data")
sampleAnnotDF %>%
  select(Monkey.ID, Timepoint, Cell.population.sorted.1) %>%
  distinct() %>%
  group_by(Timepoint, Cell.population.sorted.1) %>%
  summarize(n = n()) %>%
  ungroup() %>%
  mutate(Timepoint = factor(Timepoint, levels = c("pre", "D14", "wk8"))) %>%
  pivot_wider(names_from = Cell.population.sorted.1, values_from = n) %>%
  arrange(Timepoint) %>%
  kable()
```
Transcriptomic profiles of 12 animals for 11 subsets at 3 timepoints was performed.  
  
  
Read gene annotation
```{r read-feature-annotation, eval=FALSE}
colNames <- c("seqname",
              "source",
              "feature",
              "start",
              "end",
              "score",
              "strand",
              "frame",
              "attribute")
colTypes <- paste(c(rep("c", times = 3),
                    rep("i", times = 2),
                    rep("c", times = 4)),
                  collapse = "")
featuresAnnotationFile <- file.path(workDir, "input/joana2.genes.gtf")
skipNb <- read_lines(file = featuresAnnotationFile)
skipNb <- sum(grepl(pattern = "^#", skipNb))

featuresAnnot <- read_tsv(file      = featuresAnnotationFile,
                          skip      = skipNb,
                          col_names = colNames,
                          col_types = colTypes)
# extract gene_id and transcript_id from attributes
featAnnot <- featuresAnnot %>%
  mutate(gene_id = gsub(pattern = ".*gene_id \"([^;]+)\";.+",
                        replacement = "\\1",
                        attribute),
         gene_name = ifelse(test = grepl(pattern = "gene_name",
                                         attribute),
                            yes = gsub(pattern = ".+gene_name \"([^;]+)\";.+",
                                       replacement = "\\1",
                                       attribute),
                            no  = NA),
         gene_biotype = ifelse(test = grepl(pattern = "gene_biotype",
                                         attribute),
                            yes = gsub(pattern = ".+gene_biotype \"([^;]+)\";.*",
                                       replacement = "\\1",
                                       attribute),
                            no  = NA)) %>%
  select(seqname, strand, gene_id, gene_name, gene_biotype) %>%
  distinct() %>%
  as.data.frame()
rownames(featAnnot) <- featAnnot$gene_id
featAnnot <- featAnnot[rownames(countDF), ]
```

```{r print-feature-annotation}
cat("Number of genes annotated")
featAnnot %>%
    group_by(gene_biotype) %>%
    summarize(n = n()) %>%
    arrange(desc(n)) %>%
  kable()
```

Build raw SeqExpressionSet
```{r build-raw-seqset, eval=FALSE}
# build  raw ExpressionSet
esetRaw <- newSeqExpressionSet(counts      = as.matrix(countDF),
                               featureData = AnnotatedDataFrame(featAnnot),
                               phenoData   = AnnotatedDataFrame(sampleAnnotDF))
save(esetRaw, file = file.path(workDir, "output/joana2.esetRaw.RData"))
print(esetRaw)
```

```{r print-raw-seqset, echo=FALSE}
print(esetRaw)
```

Barplot with read alignment statistics
```{r set2-qc-align-stat, fig.width=8, fig.height=4, fig.align="left", warning=FALSE, message=FALSE}
statDF <- read_tsv(file      = file.path(workDir, "input/joana2.ReadStats.txt"),
                   col_names = FALSE) %>%
  distinct() %>%
  spread(X3, X2)
plotDF <- counts(esetRaw) %>% 
  colSums() %>%
  data.frame(Counted = .) %>%
  rownames_to_column() %>%
  merge(x = statDF, by.x = "X1", by.y = "rowname")

plotDF <- plotDF %>%
  mutate(Trimmed    = TotalReads - Surviving,
         NotMapped  = Surviving - UniqMapped - Multimapped,
         NotCounted = UniqMapped + Multimapped - Counted) %>%
  select(Trimmed, NotMapped, NotCounted, Counted, X1) %>%
  arrange(Counted) %>% 
  mutate(X1 = factor(X1, levels = unique(X1))) %>%
  gather(key, value, -X1)
ggplot(data = plotDF,
       mapping = aes(x = X1, y = value, fill = key)) +
  geom_bar(stat = "identity") +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "Samples", y = "Number of reads") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

message("number of samples with more than 2e06 counted reads")
plotDF %>%
  filter(key %in% "Counted") %>%
  mutate(over10e6 = value >= 2e06) %>%
  group_by(over10e6) %>%
  summarize(n())

message("6 samples below requirment for DEG analysis:")
pData(esetRaw)[as.vector(filter(plotDF, key %in% "Counted" & value < 2e06)$X1), ] %>%
  kable()
```
All samples except 6 have more than 2 million reads and can be used for diff.
expression analysis.
<!--
DGA7_D14_nonfCD8: total number of reads low (ReadStats: 1,150,059)
DGA7_pre_CD8: alignment to non-annotated region (align.stat: _no_feature 8366919)
DGG6_pre_CD8: unaligned reads (starLog.final.out: % of reads unmapped: too short 85.48%)
DGG6_wk8_CD8NA: alignment to non-annotated region (align.stat: _no_feature 4704602)
DGH1_wk8_CD8NA: alignment to non-annotated region (align.stat: _no_feature 4851185)
DHBC_D14_nonfCD8: 
-->  

Barplot with mapped reads statistics
```{r set2-qc-mRNA, fig.height=4, fig.width=8, fig.align="left"}
plotDF <- counts(esetRaw) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(y    = select(fData(esetRaw), gene_id, gene_biotype),
        by.x = "rowname",
        by.y = "gene_id") %>%
  mutate(gene_biotype = ifelse(test = grepl(pattern = "pseudogene",
                                            gene_biotype),
                               yes = "pseudogene",
                               no  = gene_biotype),
         gene_biotype = ifelse(test = gene_biotype %in% 
                                 c("miRNA", "lincRNA", "snoRNA",
                                   "sRNA", "snRNA", "vaultRNA",
                                   "3prime_overlapping_ncrna",
                                   "macro_lncRNA", "scaRNA"),
                               yes = "ncrna",
                               no = gene_biotype),
         gene_biotype = ifelse(test = gene_biotype %in% 
                                 c("antisense",
                                   "ncRNA",
                                   "Mt_rRNA",
                                   "Mt_tRNA",
                                   "protein_coding",
                                   "pseudogene",
                                   "ribozyme",
                                   "rRNA"),
                               yes = gene_biotype,
                               no  = "other")) %>%
  gather(sample, value, -rowname, -gene_biotype) %>%
  group_by(sample, gene_biotype) %>%
  summarize(eta = sum(value),
            mu  = mean(value),
            q2  = median(value)) %>%
  ungroup() %>%
  arrange(!grepl("Water", `sample`)) %>% 
  mutate(`sample` = factor(`sample`, levels = unique(`sample`)))

ggplot(data = plotDF,
       mapping = aes(x = sample, y = eta, fill = gene_biotype)) +
  geom_bar(stat = "identity") +
  scale_y_continuous(expand = c(0, 0)) +
  labs(x = "Samples", y = "Number of reads") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())
message("percentage of reads that are from protein coding")
plotDF %>% 
  group_by(sample) %>%
  summarize(n = sum(eta)) %>%
  merge(y = plotDF) %>%
  filter(gene_biotype %in% "protein_coding") %>%
  mutate(perc_prot_cod = eta/n * 100) %>%
  summarize(type = "perc_prot_coding",
            min  = min(.$perc_prot_cod, na.rm = TRUE),
            max  = max(.$perc_prot_cod, na.rm = TRUE)) 
``` 
As expected, most counted reads are mapped to protein coding genes.
  
Principal variance component analysis (raw)
```{r set2-pvca, fig.height=4, fig.width=4, fig.align="left", message=FALSE, warning=FALSE}
esetTemp <- ExpressionSet(assayData = log2(counts(esetRaw) + 0.25),
                          phenoData = AnnotatedDataFrame(data = pData(esetRaw)))

esetTemp$TotalReadsQ3 <- cut(esetTemp$TotalReads,
                             breaks         = quantile(esetTemp$TotalReads,
                                                       probs = c(0,0.33, 0.66,1)),
                             include.lowest = TRUE,
                             right          = TRUE)
esetTemp$NumberCellsQ3 <- cut(esetTemp$Number.of.Cells,
                              breaks         = c(0, 1500, 3000, 5000),
                              include.lowest = TRUE,
                              right          = TRUE)
esetTemp$lowCountedReads <- esetTemp$TotalReads < 2e6

fit <- PVCA(counts = exprs(esetTemp),
            meta = pData(esetTemp)[, c("Timepoint",
                                       "NHP.group",
                             "Monkey.ID",
                             "Cell.population.sorted.1",
                             "NumberCellsQ3",
                             "TotalReadsQ3",
                             "lowCountedReads")],
     threshold = 0.6,
     inter = FALSE)

plotDF <- data.frame(effect = as.vector(fit),
                     label  = names(fit)) %>%
  arrange(effect) %>%
  mutate(label = factor(label, levels = label))
ggplot(data = plotDF,
       mapping = aes(x = label, y = effect)) +
  geom_bar(stat = "identity") +
  labs(x = NULL, y = "Explained effect") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```
Subset, samples with low counts reads and animals are the main drivers of the raw gene-expression.
  
Multidimensional scaling plot based on raw counts
```{r set2-mds-raw, fig.width=6, fig.height=4, fig.align="left"}
rawMat <- counts(esetRaw)
mat <- rawMat[rowSums(rawMat) > 0, ]
mat <- log2(mat + 0.25)
distMat <- parallelDist(t(mat), threads = detectCores() - 1)
attributes(distMat)$Labels  <-  colnames(mat)
fit <- cmdscale(distMat, k = 2, eig = TRUE)
plotDF <- as.data.frame(fit$points)
plotDF <- plotDF %>%
    rownames_to_column() %>%
    merge(y = rownames_to_column(pData(esetRaw)),
          by = "rowname") %>%
  mutate(lowCountedReads = (TotalReads < 2e6))
ggplot(data    = plotDF,
                   mapping = aes(x     = V1,
                                 y     = V2,
				 color = Cell.population.sorted,
				 shape = lowCountedReads)) +
  geom_point(size = 4) +
  labs(x = paste0("1st dimension (",
                        round((fit$eig/sum(fit$eig))[1] * 100),
                        "%)"),
                    y = paste0("2nd dimension (",
                        round((fit$eig/sum(fit$eig))[2] * 100),
                        "%)")) +
  theme_bw()


filter(plotDF,V1 > 300)
```
Samples with low read counts are indeed the main driver of raw gene-expression.
  
Normalize (TMM) count data
```{r normalize-counts, message=FALSE, eval=FALSE}
# remove low counts samples
esetTemp <- esetRaw[, esetRaw$TotalReads >= 9e5]
dge <- DGEList(counts       = counts(esetTemp),
               remove.zeros = TRUE)
# remove low expressed genes
flag <- filterByExpr(dge, group = esetRaw$Cell.population.sorted)
dge <- dge[flag, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(object = dge, method = "TMM")
normalizedCounts <- cpm(dge, normalized.lib.sizes = TRUE, log = TRUE)
eset <-  newSeqExpressionSet(
    counts           = dge$counts,
    normalizedCounts = as.matrix(normalizedCounts),
    featureData      = fData(esetTemp)[rownames(normalizedCounts), ],
    phenoData        = pData(esetTemp))
save(eset, file = file.path(workDir, "output/joana2.eset.RData"))
print(eset)
```

```{r load-counts, echo=FALSE}
load(file = file.path(workDir, "output/joana2.eset.RData"))
print(eset)
```

Principal variance component analysis (normalized)
```{r set2-pvca-norm, fig.height=4, fig.width=4, fig.align="left", warning=FALSE, message=FALSE}
esetTemp <- ExpressionSet(assayData = normCounts(eset),
                          phenoData = AnnotatedDataFrame(data = pData(eset)))
esetTemp$TotalReadsQ3 <- cut(esetTemp$TotalReads,
                             breaks         = quantile(esetTemp$TotalReads,
                                                       probs = c(0,0.33, 0.66,1)),
                         include.lowest = TRUE,
                         right          = TRUE)
fit <- PVCA(counts = exprs(esetTemp),
            meta = pData(esetTemp)[, c("Timepoint",
                                       "NHP.group",
                             "Monkey.ID",
                             "Cell.population.sorted.1",
                             "TotalReadsQ3")],
                           threshold = 0.6,
                           inter = FALSE)
plotDF <- data.frame(effect = as.vector(fit),
                     label  = names(fit)) %>%
  arrange(effect) %>%
  mutate(label = factor(label, levels = unique(label)))
ggplot(data = plotDF,
       mapping = aes(x = label, y = effect)) +
  geom_bar(stat = "identity") +
  labs(x = NULL, y = "Explained effect") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```
After normalization (for total number of reads counted per samples),
subset remain the main drivers of expression.
  
Multidimensional scalling plot on normalized counts
```{r set2-mds-norm, fig.width=6, fig.height=4, fig.align="left"}
# mds normalized/cpm counts
mat <- normCounts(eset)
distMat <- parallelDist(t(mat), threads = detectCores() - 1)
attributes(distMat)$Labels  <- colnames(mat)
fit <- cmdscale(distMat, k = 2, eig = TRUE)
plotDF <- as.data.frame(fit$points) %>%
  rownames_to_column() %>%
  merge(y = rownames_to_column(pData(eset)),
        by = "rowname")
plotLabel <- round((fit$eig/sum(fit$eig)) * 100)
plotLabel <- plotLabel[1:2]
plotLabel <- paste0(c("1st dimension (", "2nd dimension ("),
                    plotLabel,
                    "%)")
ggplot(data = plotDF,
       mapping = aes(x = V1,
                     y = V2,
                     color = Cell.population.sorted.1,
                     shape = NHP.group)) +
  geom_point(size = 4) +
  labs(x = plotLabel[[1]],
       y = plotLabel[[2]]) +
  theme_bw() +
  theme(legend.key = element_blank(),
        plot.title = element_text(hjust = 0.5))
```
After normalization (for total number of reads counted per samples),
subset remain the main driver of expression

Jitter plot of genes coding for cell surface markers
```{r set2-jitter-sp, warning=FALSE, fig.width=4, fig.height=6, fig.align="left"}
geneLS <- c("CD4", "CD8A")
plotDF <- counts(eset[fData(eset)$gene_name %in% geneLS, ]) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(y = select(fData(eset), gene_id, gene_name),
	by.x = "rowname",
	by.y = "gene_id") %>%
  select(-rowname) %>%
  gather(SampleID, count, -gene_name) %>%
  merge(y = select(rownames_to_column(pData(eset)),
		   rowname,
		   Cell.population.sorted.1),
	by.x = "SampleID",
	by.y = "rowname") %>%
  mutate(value = pmax(count, 0.1))
ggplot(plotDF,
		  mapping = aes(x = Cell.population.sorted.1, y = value)) +
  geom_boxplot(outlier.color = "transparent") +
  geom_beeswarm(mapping = aes(color = Cell.population.sorted.1)) +
  scale_y_log10() +
  facet_wrap(facet = ~gene_name, scale = "free", nrow = 3) +
  theme_bw() +
  theme(axis.text  = element_text(size = 5),
	legend.pos = "none")
```

```{r set2-jitter-sp-cd4, warning=FALSE, fig.width=4, fig.height=6, fig.align="left"}
geneLS <- c("CD28", "CXCR5", "PDCD1")
plotDF <- counts(eset[fData(eset)$gene_name %in% geneLS, eset$Cell.population.sorted.1 %in%
                        c("CD4", "CD4MEM", "CD4NA", "GCTfh", "nonTfh")]) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(y = select(fData(eset), gene_id, gene_name),
	by.x = "rowname",
	by.y = "gene_id") %>%
  select(-rowname) %>%
  gather(SampleID, count, -gene_name) %>%
  merge(y = select(rownames_to_column(pData(eset)),
		   rowname,
		   Cell.population.sorted.1),
	by.x = "SampleID",
	by.y = "rowname") %>%
  mutate(value = pmax(count, 0.1))
ggplot(plotDF,
		  mapping = aes(x = Cell.population.sorted.1, y = value)) +
  geom_boxplot(outlier.color = "transparent") +
  geom_beeswarm(mapping = aes(color = Cell.population.sorted.1)) +
  scale_y_log10() +
  facet_wrap(facet = ~gene_name, scale = "free", nrow = 3) +
  theme_bw() +
  theme(axis.text  = element_text(size = 5),
	legend.pos = "none")
```
<!-- FAS/CD95 is not annotated in Mmul_10, RefSeq has a FAS annotation,
     mapped to transcript STAMBPL1-206-208 (chr9:89235316-89305989)
     while STAMBPL1 is really transcript STAMBPL1-201-205 (chr9:89096395-89234286)
-->

```{r set2-jitter-sp-cd8, warning=FALSE, fig.width=4, fig.height=6, fig.align="left"}
geneLS <- c("CCR7", "CD28", "CXCR5", "PDCD1")
plotDF <- counts(eset[fData(eset)$gene_name %in% geneLS, eset$Cell.population.sorted.1 %in%
                        c("CD8", "CD8CM", "CD8EM", "CD8NA", "fCD8", "nonfCD8")]) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(y = select(fData(eset), gene_id, gene_name),
	by.x = "rowname",
	by.y = "gene_id") %>%
  select(-rowname) %>%
  gather(SampleID, count, -gene_name) %>%
  merge(y = select(rownames_to_column(pData(eset)),
		   rowname,
		   Cell.population.sorted.1),
	by.x = "SampleID",
	by.y = "rowname") %>%
  mutate(value = pmax(count, 0.1))
ggplot(plotDF,
		  mapping = aes(x = Cell.population.sorted.1, y = value)) +
  geom_boxplot(outlier.color = "transparent") +
  geom_beeswarm(mapping = aes(color = Cell.population.sorted.1)) +
  scale_y_log10() +
  facet_wrap(facet = ~gene_name, scale = "free", nrow = 3) +
  theme_bw() +
  theme(axis.text  = element_text(size = 5),
	legend.pos = "none")
```

Principal variance component analysis per subset
```{r set2-pvca-norm-subset, fig.height=4, fig.width=4, fig.align="left", warning=FALSE, message=FALSE}
plotDF <- NULL
for (SUBSET in unique(eset$Cell.population.sorted.1)) {
  esetTemp <- eset[, eset$Cell.population.sorted.1 %in% SUBSET]
  esetTemp <- ExpressionSet(assayData = normCounts(esetTemp),
			    phenoData = AnnotatedDataFrame(data = pData(esetTemp)))
  fit <- PVCA(counts = exprs(esetTemp),
	      meta = pData(esetTemp)[, c("Timepoint",
					 "NHP.group",
					 "Monkey.ID")],
                           threshold = 0.6,
                           inter = FALSE)
  plotTemp <- data.frame(effect = as.vector(fit),
			 label  = names(fit)) %>%
    arrange(effect) %>%
    mutate(label = factor(label, levels = unique(label)),
	   Cell.population.sorted.1 = SUBSET)
  plotDF <- rbind(plotDF, plotTemp)
}

ggplot(data = plotDF,
       mapping = aes(x = label, y = effect)) +
  geom_bar(stat = "identity") +
  labs(x = NULL, y = "Explained effect") +
  theme_bw() +
  facet_wrap(facets = ~Cell.population.sorted.1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```
Donor effect is observed in all subsets
  
Substrack pre expression
```{r baselined-expression, eval=FALSE}
# merge replicate samples
flagDF <- pData(eset) %>%
  rownames_to_column() %>%
  select(Monkey.ID, Timepoint, Cell.population.sorted.1, rowname) %>%
  pivot_wider(names_from = Timepoint, values_from = rowname) %>%
  pivot_longer(cols = c("D14", "wk8"),
               names_to = "Timepoint",
               values_to = "rowname") %>%
  filter(!is.na(pre) & !is.na(rowname))
esetBaselined <- eset[, flagDF$rowname]
baselinedMat <- normCounts(eset)[, flagDF$rowname] -
  normCounts(eset)[, flagDF$pre]
normCounts(esetBaselined) <- baselinedMat
save(esetBaselined, file = file.path(workDir, "output/joana2.esetBaselined.RData"))
print(esetBaselined)
```

```{r load-baselined-expression, echo=FALSE}
load(file = file.path(workDir, "output/joana2.esetBaselined.RData"))
print(esetBaselined)
```

Principal variance component analysis per subset (on baselined expression)
```{r set2-pvca-norm-subset-baselined, fig.height=4, fig.width=4, fig.align="left", warning=FALSE, message=FALSE}
plotDF <- NULL
for (SUBSET in unique(esetBaselined$Cell.population.sorted.1)) {
  esetTemp <- esetBaselined[, esetBaselined$Cell.population.sorted.1 %in% SUBSET]
  esetTemp <- ExpressionSet(assayData = normCounts(esetTemp),
                            phenoData = AnnotatedDataFrame(data = pData(esetTemp)))
  fit <- PVCA(counts = exprs(esetTemp),
              meta = pData(esetTemp)[, c("Timepoint",
                                         "NHP.group",
                                         "Monkey.ID")],
                           threshold = 0.6,
                           inter = FALSE)
  plotTemp <- data.frame(effect = as.vector(fit),
                         label  = names(fit)) %>%
    arrange(effect) %>%
    mutate(label = factor(label, levels = unique(label)),
           Cell.population.sorted.1 = SUBSET)
  plotDF <- rbind(plotDF, plotTemp)
}

ggplot(data = plotDF,
       mapping = aes(x = label, y = effect)) +
  geom_bar(stat = "identity") +
  labs(x = NULL, y = "Explained effect") +
  theme_bw() +
  facet_wrap(facets = ~Cell.population.sorted.1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```
Treatment effect is better observed on pre-post expression. 
  
Multidimensional scalling plot on baselined counts by subsets
```{r set2-mds-baselined-subset, fig.width=8, fig.height=4.75, fig.align="left"}
plotDF <- NULL
for (SUBSET in unique(esetBaselined$Cell.population.sorted.1)) {
  # mds normalized/cpm counts
  esetTemp <- esetBaselined[, esetBaselined$Cell.population.sorted.1 %in% SUBSET]

  mat <- normCounts(esetTemp)
  distMat <- parallelDist(t(mat), threads = detectCores() - 1)
  attributes(distMat)$Labels  <- colnames(mat)
  fit <- cmdscale(distMat, k = 2, eig = TRUE)
  as.data.frame(fit$points) %>%
    rownames_to_column() %>%
    merge(y = rownames_to_column(pData(eset)),
          by = "rowname") %>%
    rbind(plotDF, .) -> plotDF
}

ggplot(data = plotDF,
                   mapping = aes(x = V1,
                                 y = V2,
                                 color = Timepoint,
                                 shape = NHP.group)) +
  geom_point(size = 4) +
  facet_wrap(facets = ~Cell.population.sorted.1, scale = "free") +
  theme_bw() +
  theme(legend.key = element_blank(),
        plot.title = element_text(hjust = 0.5))
```

Differential expression between timepoints (paired)
```{r diff-expression, eval=FALSE}
fits <- list()
# vaccine effect
for (SUBSET in unique(eset$Cell.population.sorted.1)) {
  esetTemp <- eset[, eset$Cell.population.sorted.1 %in% SUBSET]
  dge <- DGEList(counts  = counts(esetTemp),
                 samples = pData(esetTemp),
                 genes   = fData(esetTemp))
  goi <- interaction(gsub(pattern     = " .+",
                          replacement = "",
                          esetTemp$NHP.group),
                     esetTemp$Timepoint,
                     drop = TRUE)
  donor <- esetTemp$Monkey.ID
  designMat <- model.matrix(~0+goi) 
  rownames(designMat) <- sampleNames(esetTemp)
  colnames(designMat) <- gsub(pattern     = "goi",
                              replacement = "",
                              colnames(designMat))
  attr(designMat, "assign") <- attr(designMat, "contrasts") <- NULL
  v <- voom(counts = dge, design = designMat)
  corfit <- duplicateCorrelation(v$E,
                                 design = designMat,
                                 block  = donor)
  fit <- lmFit(v,
               design      = designMat,
               block       = donor,
               correlation = corfit$consensus)
  contrastLS <- grep(pattern = "pre", colnames(fit), value = TRUE, invert = TRUE) %>%
    paste0(., "-", gsub(pattern = "\\..+", replacement = ".pre", .))
  contrastMat <- makeContrasts(contrasts = contrastLS, levels = fit$design)
  fit2 <- contrasts.fit(fit = fit, contrasts = contrastMat)
  fit2 <- eBayes(fit = fit2)
  fits[[paste0(SUBSET, "_vax")]] <- list(fit = fit, fit2 = fit2)
}

# vaccine vs untreated
for (SUBSET in unique(esetBaselined$Cell.population.sorted.1)) {
  esetTemp <- esetBaselined[, esetBaselined$Cell.population.sorted.1 %in% SUBSET]
  goi <- interaction(gsub(pattern     = " .+",
                          replacement = "",
                          esetTemp$NHP.group),
                     esetTemp$Timepoint,
                     drop = TRUE)
  designMat <- model.matrix(~0+goi) 
  rownames(designMat) <- sampleNames(esetTemp)
  colnames(designMat) <- gsub(pattern     = "goi",
                              replacement = "",
                              colnames(designMat))
  attr(designMat, "assign") <- attr(designMat, "contrasts") <- NULL
  dge <- ExpressionSet(assayData =normCounts(esetTemp),
                 phenoData = AnnotatedDataFrame(pData(esetTemp)),
                 featureData   = AnnotatedDataFrame(fData(esetTemp)))
  fit <- lmFit(dge,
               design      = designMat)
  contrastLS <- grep(pattern = "UnTx", colnames(fit), value = TRUE, invert = TRUE) %>%
    paste0(., "-", gsub(pattern = ".+\\.", replacement = "UnTx.", .))
  contrastMat <- makeContrasts(contrasts = contrastLS, levels = fit$design)
  fit2 <- contrasts.fit(fit = fit, contrasts = contrastMat)
  fit2 <- eBayes(fit = fit2)
  fits[[paste0(SUBSET, "_untx")]] <- list(fit = fit, fit2 = fit2)
}
# save fits
save(fits, file = file.path(workDir, "output/joana2.fits.RData"))
```

```{r load-diff-expression, echo=FALSE}
load(file = file.path(workDir, "output/joana2.fits.RData"))
```

```{r print-nb-deg}
# print number of genes differently expressed
degNbDF <- NULL 
for (modelName in names(fits)) {
  fit2 <- fits[[modelName]][["fit2"]]
  for (coefName in colnames(fit2)) {
    top <- topTable(fit = fit2, coef = coefName, number = Inf) %>%
      as.data.frame() %>%
      filter(gene_biotype %in% "protein_coding") %>%
      group_by(sign(logFC)) %>%
      summarize(p     = sum(P.Value <= 0.05),
                adj.p = sum(adj.P.Val <= 0.05)) %>%
      mutate(modelName = modelName,
             contrast = coefName) %>%
      rbind(degNbDF) -> degNbDF
  }
}

degNbDF %>%
  pivot_longer(cols = c(-modelName, -contrast, -`sign(logFC)`),
               names_to = "cname",
               values_to = "n") %>%
  mutate(`sign(logFC)` = c("-1" = "DN", "1" = "UP")[as.character(`sign(logFC)`)],
         cname         = paste0(cname, `sign(logFC)`)) %>%
  select(-`sign(logFC)`) %>%
  pivot_wider(names_from = cname, values_from = n) %>%
  mutate(p = paste0(pUP, "/", pDN),
         adj.p = paste0(adj.pUP, "/", adj.pDN)) %>%
  select(modelName, contrast, p, adj.p) %>%
  as.data.frame() %>%
  kable()
```

Heatmaps with DEGs
```{r set2-heatmap-deg-top50, fig.align="left", fig.height=8.5, fig.width=6}
for (modelName in grep(pattern = "_vax", names(fits), value = TRUE)) {
  fit2 <- fits[[modelName]][["fit2"]]
  for (coefName in colnames(fit2)) {
    top <- topTable(fit = fit2, coef = coefName, number = Inf) %>%
      as.data.frame() %>%
      filter(!is.na(gene_name) &
	       gene_biotype %in% "protein_coding" &
	       adj.P.Val <= 0.05)
    if (nrow(top) > 0) {
      top <- top %>%
        top_n(50, wt = abs(t)) %>%
        arrange(desc(logFC))
      sampleLS <- which(fit2$contrast[, coefName] != 0) %>%
        fit2$design[, .] %>%
        as.data.frame() %>%
        rownames_to_column() %>%
        gather(group, value, -rowname) %>%
        group_by(rowname) %>%
        summarize(value = sum(value)) %>%
        filter(value > 0) %>%
        .$rowname
      mat <- normCounts(eset)[top$gene_id, sampleLS, drop = FALSE] %>%
        t() %>%
        scale() %>%
        t()
      mat <- mat[complete.cases(mat), , drop = FALSE]
      colorLS <- colorRampPalette(colors = c("blue", "white", "red"))(n = 100)
      breakLS <- c(-1 * max(abs(mat), na.rm = TRUE),
                   seq(from = -1 * min(abs(range(mat, na.rm = TRUE))),
                       to   = min(abs(range(mat, na.rm = TRUE))),
                       length.out = 99),
                   max(abs(mat), na.rm = TRUE))
      matAnnot <- pData(eset) %>%
        rownames_to_column() %>%
        select(rowname, Timepoint, NHP.group, Cell.population.sorted.1) %>%
        column_to_rownames()
      matAnnot <- matAnnot[colnames(mat), ]
      matAnnot <- matAnnot[, colMeans(is.na(matAnnot)) < 1]
      ha <- HeatmapAnnotation(df = matAnnot)
      heat <- Heatmap(matrix         = mat,
                      width          = unit(0.5, units = "npc"),
                      height         = unit(0.1+0.6*(nrow(mat)/50), units = "npc"),
                      top_annotation = ha,
                      name           = "zscore",
                      column_title   = paste0(modelName, ": ", coefName),
                      row_names_gp   = gpar(fontsize = 7),
                      column_labels  = pData(eset)[colnames(mat), "Monkey.ID"],
		                  row_labels     = top$gene_name,
		                  use_raster     = TRUE)
      print(heat)
    }
  }
}
```

```{r set2-heatmap-deg-top50-untx, efig.align="left", fig.height=8.5, fig.width=6}
for (modelName in grep(pattern = "_untx", names(fits), value = TRUE)) {
  fit2 <- fits[[modelName]][["fit2"]]
  for (coefName in colnames(fit2)) {
    top <- topTable(fit = fit2, coef = coefName, number = Inf) %>%
      as.data.frame() %>%
      filter(!is.na(gene_name) &
	       gene_biotype %in% "protein_coding" &
	       adj.P.Val <= 0.05)
    if (nrow(top) > 0) {
      top <- top %>%
        top_n(50, wt = abs(t)) %>%
        arrange(desc(logFC))
      sampleLS <- which(fit2$contrast[, coefName] != 0) %>%
        fit2$design[, .] %>%
        as.data.frame() %>%
        rownames_to_column() %>%
        gather(group, value, -rowname) %>%
        group_by(rowname) %>%
        summarize(value = sum(value)) %>%
        filter(value > 0) %>%
        .$rowname
      mat <- normCounts(esetBaselined)[top$gene_id, sampleLS, drop = FALSE]
      mat <- mat[complete.cases(mat), , drop = FALSE]
      colorLS <- colorRampPalette(colors = c("blue", "white", "red"))(n = 100)
      breakLS <- c(-1 * max(abs(mat), na.rm = TRUE),
                   seq(from = -1 * min(abs(range(mat, na.rm = TRUE))),
                       to   = min(abs(range(mat, na.rm = TRUE))),
                       length.out = 99),
                   max(abs(mat), na.rm = TRUE))
      matAnnot <- pData(esetBaselined) %>%
        rownames_to_column() %>%
        select(rowname, Timepoint, NHP.group, Cell.population.sorted.1) %>%
        column_to_rownames()
      matAnnot <- matAnnot[colnames(mat), ]
      matAnnot <- matAnnot[, colMeans(is.na(matAnnot)) < 1]
      ha <- HeatmapAnnotation(df = matAnnot)
      heat <- Heatmap(matrix         = mat,
                      width          = unit(0.5, units = "npc"),
                      height         = unit(0.1+0.6*(nrow(mat)/50), units = "npc"),
                      top_annotation = ha,
                      name           = "zscore",
                      column_title   = paste0(modelName, ": ", coefName),
                      row_names_gp   = gpar(fontsize = 7),
                      column_labels  = pData(eset)[colnames(mat), "Monkey.ID"],
		                  row_labels     = top$gene_name,
		                  use_raster     = TRUE)
      print(heat)
    }
  }
}
```

Heatmaps with all DEGs across time by subset
```{r set2-deg-heat-f, fig.align="left", fig.height=8.5, fig.width=6, message=FALSE}
message("FDR cutoff of 5%") 
for (modelName in grep(pattern = "vax", names(fits), value = TRUE)) {
  fit2 <- fits[[modelName]][["fit2"]]
  top <- topTable(fit2, number = Inf, p.value = 0.05) %>%
    as.data.frame() %>%
    filter(gene_biotype %in% "protein_coding")
  if (nrow(top) > 0) {
    top <- top %>%
      arrange(desc(WT.D14.WT.pre)) %>%
      mutate(gene_name = ifelse(test = is.na(gene_name),
				yes  = gene_id,
				no   = gene_name))
    sampleLS <- rownames(fit2$design)
    sampleLS <- intersect(sampleLS, sampleNames(esetBaselined))
    mat <- normCounts(esetBaselined)[top$gene_id, sampleLS, drop = FALSE]
      
    mat <- mat[complete.cases(mat), ]
    matAnnot <- pData(eset) %>%
      rownames_to_column() %>%
      select(rowname, Timepoint, NHP.group, Cell.population.sorted.1) %>%
      column_to_rownames()
    matAnnot <- matAnnot[colnames(mat), ]
    ha <- HeatmapAnnotation(df = matAnnot,
                            col = list(NHP.group = c("UnTx" = "grey", "WT bNAb Tx" = "orange", "DEL bNAb Tx" = "green")))
    splitCols <- pData(eset)[colnames(mat),
			     c("Timepoint", "NHP.group")] %>%
      apply(MARGIN = 1, FUN = paste, collapse = ".") 
    heat <- Heatmap(matrix         = mat,
		    width          = unit(0.5, units = "npc"),
		    height         = unit(0.75, units = "npc"),
		    top_annotation = ha,
		    name           = "log2FC",
		    column_title   = modelName,
		    show_row_names = FALSE,
		    column_labels  = pData(eset)[colnames(mat), "Monkey.ID"],
		    column_split   = splitCols,
		    use_raster     = TRUE)
    print(heat)
  }
}
```

```{r set2-deg-heat-f-untx, fig.align="left", fig.height=8.5, fig.width=6, message=FALSE}
message("FDR cutoff of 5%") 
for (modelName in grep(pattern = "untx", names(fits), value = TRUE)) {
  fit2 <- fits[[modelName]][["fit2"]]
  top <- topTable(fit2, number = Inf, p.value = 0.05) %>%
    as.data.frame()
  if (nrow(top) > 0) { 
    top <- top %>%
    filter(gene_biotype %in% "protein_coding")
  }
  if (nrow(top) > 0) {
    top <- top %>%
      arrange(desc(WT.D14.UnTx.D14)) %>%
      mutate(gene_name = ifelse(test = is.na(gene_name),
				yes  = gene_id,
				no   = gene_name))
    sampleLS <- rownames(fit2$design)
    sampleLS <- intersect(sampleLS, sampleNames(esetBaselined))
    mat <- normCounts(esetBaselined)[top$gene_id, sampleLS, drop = FALSE]
      
    mat <- mat[complete.cases(mat), ]
    matAnnot <- pData(eset) %>%
      rownames_to_column() %>%
      select(rowname, Timepoint, NHP.group, Cell.population.sorted.1) %>%
      column_to_rownames()
    matAnnot <- matAnnot[colnames(mat), ]
    ha <- HeatmapAnnotation(df = matAnnot,
                            col = list(NHP.group = c("UnTx" = "grey", "WT bNAb Tx" = "orange", "DEL bNAb Tx" = "green")))
    splitCols <- pData(eset)[colnames(mat),
			     c("Timepoint", "NHP.group")] %>%
      apply(MARGIN = 1, FUN = paste, collapse = ".") 
    heat <- Heatmap(matrix         = mat,
		    width          = unit(0.5, units = "npc"),
		    height         = unit(0.75, units = "npc"),
		    top_annotation = ha,
		    name           = "log2FC",
		    column_title   = modelName,
		    row_labels     = top$gene_name,
		    column_labels  = pData(eset)[colnames(mat),
						 "Monkey.ID"],
		    column_split   = splitCols)
    print(heat)
  }
}
```

```{r session-info}
sessionInfo()
```
